### PR TITLE
Fixed #21381 -- Remove contrib.redirects dependency on contrib.sites

### PR DESCRIPTION
**Pull request by dspechnikov**
Redirect model now has optional `domain` field. If it's specified and matched with request domain, such redirect will be used over the one with the same `old_path` and blank domain.

Other than that, redirect logic is the same.

If `sites` framework is installed, admin redirect form will have optional `site` field which can be used to fill redirect domain.

For existing database installations, existing redirect sites domains data will be migrated to new redirect domain field.

TODO:

- [x] Raise an exception in data migration if sites framework is not installed to prevent data loss
- [x] <del>via @knbk: maybe we could add the necessary migrations for the model changes including the data migration, and include a squashed migration without the sites dependency so that new deployments don't need to install sites just to migrate</del> add `site` foreign key conditionally instead. If any redirects data exist, require `sites` app to be installed to prevent data loss. New deployments don't need `sites` app at all.
- [x] rewrite form save
- [x] <del>consider adding reverse migration</del> I think it doesn't make sense here since it would require sites app installed again in order to redirects app to function.
- [x] add sites app requirement for existing data migration to docs

### CODE REVIEW

**Comment by dspechnikov** [comment]
Man, that problem with tests described in [6bb4797](https://github.com/django/django/pull/9023/commits/6bb479768276fc37d97227373768d254c997308c) was really hard to track.

Either I don't understand something or the whole test system needs to be refactored to make tests more isolated.

File: django/contrib/redirects/migrations/0002_remove_site.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+            name='domain',
+            field=models.CharField(
+                blank=True,
+                help_text='If set, redirect requests from this domain only.',
+                max_length=255,
+                verbose_name='domain'
+            ),
+        ),
+
+        migrations.RunPython(migrate_site_domains, migrations.RunPython.noop),
```
Comment (knbk) [inline django/contrib/redirects/migrations/0002_remove_site.py]: This operation needs a separate migration. Since Postgres executes migrations in a transaction, but doesn't allow both schema changes and data changes in a single migration, this will fail with `cannot ALTER TABLE "django_redirect" because it has pending trigger events` if there is any existing data.

Comment (dspechnikov) [reply django/contrib/redirects/migrations/0002_remove_site.py]: Thanks, I forgot about differences between sqlite and psql. Fixed it.

Any other concerns?

File: django/contrib/redirects/migrations/0001_initial.py [hunk 2]
```diff
@@ -1,36 +1,43 @@
+# -*- coding: utf-8 -*-
+from __future__ import unicode_literals
+
```
Comment (rsalmaso) [inline django/contrib/redirects/migrations/0001_initial.py]: python 2 code here

Comment (dspechnikov) [reply django/contrib/redirects/migrations/0001_initial.py]: fixed, thanks

**Comment by knbk** [comment]
If people attempt to remove the sites app after migrating by executing `python manage.py sites zero` before removing the app from `INSTALLED_APPS`, they will lose all data in the redirects table. Since this is the usual procedure for uninstalling apps, the docs should not say "After that you can safely uninstall it." without any warning. 

I've just tested the approach with squashed migrations and realised that the dependencies on the sites app are automatically added to the squashed migration. I'm currently working on a patch that fixes that. 

If that patch doesn't work out, I still think a squashed migration is the correct approach. In this case, we can consider the time before the replaced migrations are removed as a deprecation period in which the sites app is still required. After that period we can remove the replaced migrations, and with it, the dependency on the sites app.

**Comment by dspechnikov** [comment]
Thanks for pointing this out, I totally missed the dependencies issue.

Squashed migration could definitely help here. Waiting for your patch, it seems like just removing sites migrations from dependencies should do it.

If all else fails, I have some crazy idea about adding site field in RunPython code. I'm not even sure it will work, but still 😅

**Comment by dspechnikov** [comment]
I played a little with squashed migration option and think it is not really viable solution. Here's why:
  1. Squashed migration always depends on `sites` migrations if replaced migrations exist since they depend on it. 
  2. All deployments which are not updated to the latest replaced migration will break if remove those replaced migrations. And we can't be sure that there are no such deployments.

Besides, deprecation period for migrations just looks awkward in my opinion :/

So I thought about alternatives keeping in mind these points:
- There will be 2 types of deployments at the moment this change will be introduced - with sites data in `redirects` table and clean ones.
- Neither of them should depend on `sites` framework in any aspect.
- `sites` references data must be migrated.
- It's nice to be able to do `migrate redirects zero`.

Considering all that, I came to conclusion (and tested) that the initial solution with `site_id` IntegerField satisfies all of them. Even that wouldn't be needed, but some db backends like sqlite3 add new fields by copying all fields which are in project state. Still, I don't think it'll cause any troubles. 

@knbk Please comment if you have any concerns or objections.

**Comment by knbk** [comment]
The first issue is exactly what I want to fix. The patch is almost finished, I just got caught up in other things and didn't get around to creating a ticket and finishing the patch. With that patch applied, there won't be any dependency on the sites framework as soon as all (or none) of the replaced migrations are applied. Existing deployments must migrate before removing the sites app. 

I'm quite busy at the moment, and I'm not sure when I'll have time to finish the patch. It won't be ready for the feature freeze for 2.0 this Monday, but I don't think we should rush this.

**Comment by dspechnikov** [comment]
Agree, it can easily wait till 2.1. I just don't want the PR to hang here forever. Contact me if you need any help with the patch.

BTW, I'm really interested to see a way to workaround replaced migrations dependencies.

File: django/contrib/redirects/forms.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+        super().__init__(*args, **kwargs)
+
+        if apps.is_installed('django.contrib.sites'):
+            from django.contrib.sites.models import Site
+
+            self.fields['site'] = forms.ModelChoiceField(
+                queryset=Site.objects.all(),
+                to_field_name='domain',
+                empty_label='Any',
+                widget=admin.widgets.AdminRadioSelect(attrs={
```
Comment (codingjoe) [inline django/contrib/redirects/forms.py]: This will only work for up to 3-5 sites. I would recommend using a select input.

File: django/contrib/redirects/forms.py [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+            instance.domain = getattr(site, 'domain', '')
+
+        if commit:
+            instance.save()
+
+        return instance
+
+    class Meta:
+        model = Redirect
+        fields = '__all__'
```
Comment (codingjoe) [inline django/contrib/redirects/forms.py]: `site` is a field on `Redirect` therefore it will be displayed even if `django.contrib.sites` is not installed.

File: django/contrib/redirects/forms.py [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+                }),
+                label=_('Site'),
+                help_text=_('If domain is not set, redirect from this site.'),
+                required=False,
+            )
+
+    def save(self, commit=True):
+        instance = super().save(commit=False)
+
+        if apps.is_installed('django.contrib.sites') and not instance.domain:
```
Comment (codingjoe) [inline django/contrib/redirects/forms.py]: I would recommend putting this inside a `clean` function, not in save.

File: django/contrib/redirects/forms.py [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+            site = self.cleaned_data.get('site')
+
+            instance.domain = getattr(site, 'domain', '')
```
Comment (codingjoe) [inline django/contrib/redirects/forms.py]: You are silently overwriting a value displayed in a form. You should exclude it entirely form the form or set it dynamically using JavaScript.

File: django/contrib/redirects/forms.py (hunk 4 above)
Comment (codingjoe) [reply django/contrib/redirects/forms.py]: You could also just raise a `ValidationError` should the site's domain and the entered domain not match or make the exclusive.

File: django/contrib/redirects/middleware.py [hunk 7]
```diff
@@ -1,50 +1,39 @@
...
+
+        r = Redirect.objects.filter(
+            domain__in=[domain, ''],
+            old_path__in=paths
+        ).order_by('-domain', 'old_path').first()
+
         if r is not None:
             if r.new_path == '':
                 return self.response_gone_class()
+
```
Comment (codingjoe) [inline django/contrib/redirects/middleware.py]: unrelated

File: django/contrib/redirects/middleware.py [hunk 8]
```diff
@@ -1,50 +1,39 @@
...
             return self.response_redirect_class(r.new_path)
 
-        # No redirect was found. Return the response.
```
Comment (codingjoe) [inline django/contrib/redirects/middleware.py]: why did you remove the commend? Is it no longer valid?

File: django/contrib/redirects/migrations/0001_initial.py [hunk 9]
```diff
@@ -3,21 +3,14 @@
...
 
-    dependencies = [
-        ('sites', '0001_initial'),
-    ]
+    dependencies = []
 
     operations = [
         migrations.CreateModel(
             name='Redirect',
+
```
Comment (codingjoe) [inline django/contrib/redirects/migrations/0001_initial.py]: unrelated

File: django/contrib/redirects/migrations/0001_initial.py [hunk 10]
```diff
@@ -27,10 +20,15 @@ class Migration(migrations.Migration):
                     help_text="This can be either an absolute path (as above) or a full URL starting with 'http://'.",
                     max_length=200, verbose_name='redirect to', blank=True
                 )),
+                # We need to add `domain` field to migrate sites references
+                # data. Some db backends add fields by creating new table
+                # and copying all fields which are in project state. This
+                # field helps preserve `sites` references without requiring
+                # dependency on 'django.contrib.sites' migrations.
+                ('site_id', models.IntegerField(null=True)),
```
Comment (codingjoe) [inline django/contrib/redirects/migrations/0001_initial.py]: I am not so sure about that change. This is not an integer field, but a FK. It's not nullable, it has an index and some databases have other pseudo tables to reference a FK field.
I am not sure if this actually works using all Databases. I would recommend to be extra careful when changing old migrations.

File: django/contrib/redirects/migrations/0001_initial.py [hunk 11]
```diff
@@ -27,10 +20,15 @@ class Migration(migrations.Migration):
...
             ],
             options={
                 'ordering': ('old_path',),
-                'unique_together': {('site', 'old_path')},
```
Comment (codingjoe) [inline django/contrib/redirects/migrations/0001_initial.py]: Same here, the unique constraint is only removed in the ORM but will still persist in the database.

File: django/contrib/redirects/migrations/0002_add_domain.py [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+
+
+def check_redirects_data(apps, schema_editor):
+    db_alias = schema_editor.connection.alias
+
+    Redirect = apps.get_model('redirects', 'Redirect')
+
+    if (Redirect.objects.using(db_alias).exists() and
+            not django_apps.is_installed('django.contrib.sites')):
+        raise ImproperlyConfigured(
```
Comment (codingjoe) [inline django/contrib/redirects/migrations/0002_add_domain.py]: Why does it need to be installed in this migration? This migration doesn't even do anything. It's not given that all new migrations are applied at the same time.

File: django/contrib/redirects/migrations/0003_migrate_sites.py [hunk 13]
```diff
@@ -0,0 +1,31 @@
+from django.db import migrations
+
+
+def migrate_site_domains(apps, schema_editor):
```
Comment (codingjoe) [inline django/contrib/redirects/migrations/0003_migrate_sites.py]: I would suggest to merge this and the `0002` migration into one.

File: django/contrib/redirects/migrations/0003_migrate_sites.py [hunk 14]
```diff
@@ -0,0 +1,31 @@
...
+    try:
+        Site = apps.get_model('sites', 'Site')
+    except LookupError:
+        return
+
+    Redirect = apps.get_model('redirects', 'Redirect')
+
+    db_alias = schema_editor.connection.alias
+    for r in Redirect.objects.using(db_alias).all():
```
Comment (codingjoe) [inline django/contrib/redirects/migrations/0003_migrate_sites.py]: This migration will be very slow. I would recommend using the `F` object to perform the migration in a single update statement rather than looping over all rows.

File: django/contrib/redirects/migrations/0004_remove_site.py [hunk 15]
```diff
@@ -0,0 +1,19 @@
...
+
+
+class Migration(migrations.Migration):
+
+    dependencies = [
+        ('redirects', '0003_migrate_sites'),
+    ]
+
+    operations = [
+        migrations.AlterUniqueTogether(
```
Comment (codingjoe) [inline django/contrib/redirects/migrations/0004_remove_site.py]: This is a long list of migrations. I would recommend to try to get as many operations into one migration as you can. Loading a lot of migrations will slow down Django during startup and all migration commands.

File: docs/ref/contrib/redirects.txt [hunk 16]
```diff
@@ -14,28 +14,39 @@ Installation
...
-3. Add ``'django.contrib.redirects.middleware.RedirectFallbackMiddleware'``
+1. Add ``'django.contrib.redirects'`` to your :setting:`INSTALLED_APPS` setting.
+2. Add ``'django.contrib.redirects.middleware.RedirectFallbackMiddleware'``
    to your :setting:`MIDDLEWARE` setting.
-4. Run the command :djadmin:`manage.py migrate <migrate>`.
+3. Run the command :djadmin:`manage.py migrate <migrate>`.
+
+.. admonition:: Sites data migration
+
+    If you have existing redirects data with
```
Comment (codingjoe) [inline docs/ref/contrib/redirects.txt]: This is only true for a migration from an old version of Django. Therefore it should only be inside the release documentation  or in `versionchanged` note.

File: django/contrib/redirects/admin.py [hunk 17]
```diff
@@ -1,10 +1,34 @@
...
 
 
 @admin.register(Redirect)
 class RedirectAdmin(admin.ModelAdmin):
-    list_display = ('old_path', 'new_path')
-    list_filter = ('site',)
-    search_fields = ('old_path', 'new_path')
-    radio_fields = {'site': admin.VERTICAL}
+    form = RedirectForm
+
```
[the remark this answers is not in the document]
Comment (dspechnikov) [reply django/contrib/redirects/admin.py]: It's for separating form and fields configuration. If it violates some coding style guidelines which I'm not aware of, please point it out.

File: django/contrib/redirects/forms.py (hunk 3 above)
Comment (dspechnikov) [reply django/contrib/redirects/forms.py]: Select is less convenient when there are not many sites, which I believe is the vast majority of applications. Also, it was the radio select before, so nothing became worse here at least ^^

File: django/contrib/redirects/forms.py (hunk 5 above)
Comment (dspechnikov) [reply django/contrib/redirects/forms.py]: There is no form validation, so why should this be in clean?

File: django/contrib/redirects/forms.py (hunk 6 above)
Comment (dspechnikov) [reply django/contrib/redirects/forms.py]: So you're saying nobody reads help texts... Even in this case, `site` and `domain` fields have the same purpose, so I don't think it will be that confusing.

File: django/contrib/redirects/forms.py (hunk 4 above)
Comment (dspechnikov) [reply django/contrib/redirects/forms.py]: Did you check that `site` field is really displayed? It's added dynamically in form constructor and isn't displayed as soon as I remove `django.contrib.sites` from `INSTALLED_APPS`.

**Comment by dspechnikov** [comment]
@codingjoe Here are my thoughts about `sites` framework support.
- `redirects` app now has the same logic regardless of any other application installed, including `sites`. You create redirect object with paths and source domain, and middleware always handles it exactly the same way.
- Previous django versions users may become used to convenient `site` selector when it comes to creating redirect, and it would be nice to support it. Hence rather simple logic: if they have `sites` framework installed, they're able to set redirect domain via the same form control.
- It does not change core `redirects` app logic in any way, just adds some convenience. Don't think this is really that magical.

File: django/contrib/redirects/middleware.py (hunk 8 above)
Comment (dspechnikov) [reply django/contrib/redirects/middleware.py]: It's still valid, but it clutters the code which is pretty self explanatory in my opinion.

File: django/contrib/redirects/migrations/0001_initial.py (hunk 11 above)
Comment (dspechnikov) [reply django/contrib/redirects/migrations/0001_initial.py]: Since `site_id` is removed in further migration, it shouldn't be a problem.

File: django/contrib/redirects/migrations/0002_add_domain.py (hunk 12 above)
Comment (dspechnikov) [reply django/contrib/redirects/migrations/0002_add_domain.py]: It's here to prevent schema change by AddField and possible data loss. 

Technically you're correct that migrations can be applied separately, but documentation explicitly says 'Run python manage.py migrate'. Also, the app cannot be considered working until all migrations are applied anyway.

File: django/contrib/redirects/migrations/0003_migrate_sites.py (hunk 13 above)
Comment (dspechnikov) [reply django/contrib/redirects/migrations/0003_migrate_sites.py]: Schema change and data migration should be separated, see @knbk comment above.

File: django/contrib/redirects/migrations/0003_migrate_sites.py (hunk 14 above)
Comment (dspechnikov) [reply django/contrib/redirects/migrations/0003_migrate_sites.py]: As far as I understand (and tried) F object cannot be used to update from related field. If you'll be able to do it, feel free to send working example to me.

File: django/contrib/redirects/migrations/0004_remove_site.py (hunk 15 above)
Comment (dspechnikov) [reply django/contrib/redirects/migrations/0004_remove_site.py]: I'd also prefer it, but data migration needs to be separated.

File: docs/ref/contrib/redirects.txt (hunk 16 above)
Comment (dspechnikov) [reply docs/ref/contrib/redirects.txt]: Fair point, I'll update the docs, thanks.

File: django/contrib/redirects/migrations/0001_initial.py (hunk 10 above)
Comment (dspechnikov) [reply django/contrib/redirects/migrations/0001_initial.py]: Though it may be valid concern, I really would like some example where it actually doesn't work.

The only reason this field is actually there is to preserve it in project state, it shouldn't affect existing database in any way.

**Comment by dspechnikov** [comment]
@knbk any news regarding squashmigrations patch?

File: django/contrib/redirects/admin.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (codingjoe) [reply django/contrib/redirects/admin.py]: The form class is configuration too, is it not?
You can group class attributes using a blank line, but this attribute is inherited from `BaseModelAdmin` where it is not separated. It just seemed odd.

File: django/contrib/redirects/forms.py (hunk 3 above)
Comment (codingjoe) [reply django/contrib/redirects/forms.py]: hm... ok. fair enough, maybe it makes sense to make it swappable, but I don't want to overcomplicate things.

File: django/contrib/redirects/forms.py (hunk 5 above)
Comment (codingjoe) [reply django/contrib/redirects/forms.py]: `clean` is not only for validation, but also for data modification in a form.

File: django/contrib/redirects/forms.py (hunk 6 above)
Comment (codingjoe) [reply django/contrib/redirects/forms.py]: Yes I hope people do read help texts, but the help text doesn't imply the current behavior. It overwrites the domain provided in a form, no matter what.

File: django/contrib/redirects/forms.py (hunk 4 above)
Comment (codingjoe) [reply django/contrib/redirects/forms.py]: No, I have only reviewed the code on it's own, haven't tried it yet.

File: django/contrib/redirects/middleware.py (hunk 8 above)
Comment (codingjoe) [reply django/contrib/redirects/middleware.py]: I presume who ever put it there had a different opinion ;) I would recommend to leave it be.

File: django/contrib/redirects/migrations/0001_initial.py (hunk 11 above)
Comment (codingjoe) [reply django/contrib/redirects/migrations/0001_initial.py]: That implies that the constraint cascased, again I don't know if this is the case of all DBs.

File: django/contrib/redirects/migrations/0002_add_domain.py (hunk 12 above)
Comment (codingjoe) [reply django/contrib/redirects/migrations/0002_add_domain.py]: How is the schema change related to data loss?
I would recommend to have those checks in the next migration, the data migration.

File: django/contrib/redirects/migrations/0003_migrate_sites.py (hunk 13 above)
Comment (codingjoe) [reply django/contrib/redirects/migrations/0003_migrate_sites.py]: ok, fair enough

File: django/contrib/redirects/migrations/0003_migrate_sites.py (hunk 14 above)
Comment (codingjoe) [reply django/contrib/redirects/migrations/0003_migrate_sites.py]: Dang, you are right. I forgot.
Can you use the `iterator` than, to avoid migrations going out of memory?

**Comment by codingjoe** [comment]
Sorry that I didn't have the time until now. I am still not so sure about the migration part. I don't have much experience in DBs like Oracle. Switching from a FK to a nullable int field seems risky.

File: django/contrib/redirects/admin.py [hunk 18]
```diff
@@ -1,10 +1,34 @@
...
+                'fields': source_fields
+            }),
+            (_('Destination'), {
+                'fields': ('new_path',)
+            }),
+        )
+
+    def get_form(self, request, obj=None, **kwargs):
+        kwargs['fields'] = ('old_path', 'domain', 'new_path')
+
```
Comment (timgraham) [inline django/contrib/redirects/admin.py]: Generally, your coding style uses many more blank lines than the rest of Django.

File: django/contrib/redirects/migrations/0001_initial.py [hunk 19]
```diff
@@ -27,10 +20,15 @@ class Migration(migrations.Migration):
                     help_text="This can be either an absolute path (as above) or a full URL starting with 'http://'.",
                     max_length=200, verbose_name='redirect to', blank=True
                 )),
+                # We need to add `domain` field to migrate sites references
```
Comment (timgraham) [inline django/contrib/redirects/migrations/0001_initial.py]: Avoid use of "we" in comments per our [style guide](https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/coding-style/#python-style).

File: tests/redirects_tests/test_middleware.py [hunk 20]
```diff
@@ -0,0 +1,131 @@
...
+        self.rf = RequestFactory()
+
+    def test_blank_redirect(self):
+        r = Redirect.objects.create(domain='', old_path='/old', new_path='/new')
+        request = self.rf.get(r.old_path)
+
+        response = self.middleware(request)
+
+        self.assertRedirects(response, r.new_path,
+                             status_code=301, fetch_redirect_response=False)
```
Comment (timgraham) [inline tests/redirects_tests/test_middleware.py]: This should use longer lines (up to 119 characters) or hanging indent as described in [Python coding style](https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/coding-style/#python-style).

**Comment by dspechnikov** [comment]
@timgraham Do you have any ideas about what kind of backward compatible solution could solve the issue?

I don't see one, hence implemented it this way. I've put it for review in IRC channel and developers mailing list but got no other ideas except one mentioned by @knbk above.

If this kind of backward incompatibility is unacceptable, I'll close it.
